### PR TITLE
Issue/quick 258 report max action error rate

### PR DESCRIPTION
**Pull request by dagguh**
## Before
[JSW 7.2.0](https://quicksilver-jhwr-cache-ireland.s3-eu-west-1.amazonaws.com/QUICK-132-fix-v3/jira-exploration-chart.html?versionId=35dEJnwm7SQZHIUqRuiBGCv8GGfOQ4yd): image

[JSW 7.13.0](https://quicksilver-jhwr-cache-ireland.s3-eu-west-1.amazonaws.com/QUICK-240-redo-jsw-7.13.0/xl-jsw-7.13.0/jira-exploration-chart.html?versionId=9p1T8PoBODgEjiHhDlPJinGE00C5G76D): image

## After
[JSW 7.2.0](https://quicksilver-jhwr-cache-ireland.s3-eu-west-1.amazonaws.com/QUICK-132-fix-v3/jira-exploration-chart.html?versionId=Cf9U1dd4qtduErDV2VJRmKKhuF7k5K1t): image

[JSW 7.13.0](https://quicksilver-jhwr-cache-ireland.s3-eu-west-1.amazonaws.com/QUICK-240-redo-jsw-7.13.0/xl-jsw-7.13.0/jira-exploration-chart.html?versionId=LtQspjZ9D3X.dCoPmcnpkt9qS0K7Tjo2): image

### CODE REVIEW

File: src/test/resources/hardware-exploration-chart-template.html [hunk 1]
```diff
@@ -80,7 +80,10 @@
                 responsive: true,
                 title: {
                     display: true,
-                    text: '<%= application =%> hardware exploration: error rate'
+                    text: [
+                        '<%= application =%> hardware exploration: max action error rate',
+                        'how broken is the most broken action?'
```
Comment (dagguh) [inline src/test/resources/hardware-exploration-chart-template.html:134]: I changed the chart title to better explain what it means now. I didn't refresh the screenshots after that.

**Review by mminns — COMMENTED** [review]
As there still seems to be a question about which is the best measure/graph for error rate, why don’t we keep both, simply adding another graph to the output? They both have value.

Would it be possible to add the action label to the hover over dialog-thing (a techincal term ;) ) ?

Screen Shot 2019-07-26 at 3 07 09 PM

Other than that the implementation LGTM.

**Comment by dagguh** [comment]
> As there still seems to be a question about which is the best measure/graph for error rate, why don’t we keep both, simply adding another graph to the output? They both have value.

### Show what matters
Currently, we can glance at 3 charts and we can quickly understand the exploration results.
If we add the thresholds used (apdex, error rate), then the exploration process itself would become obvious. *If* we use the overall error rate in the exploration logic, then we should include it in the chart to maintain the transparency.

### Minimize cognitive load
We still should strive for fewer charts. We considered adding a chart for every action type. They would have a lot of value, but the costs are very important. Eyeballing 3 + 15 = 18 charts would no longer be quick. Psychology says that humans can hold 4 chunks of information in their operating memory. So adding the fourth chart would be ok, but the 5th would be too much.
The detailed data is also already available, e.g. in the `summary-per-cohort.html` files, so deep dives are still possible.

### Consider the tradeoff
Should the overall error rate (OER) guide the exploration?
IMHO, it should *iff* the max action error rate (MAER) is insufficient.
Current results show that the MAER threshold would need to be around 3-5%. If all actions fail on 3% level, we would miss that. OER threshold of 1% would catch that.

OER offers value when MAER has a high threshold. If we lower the threshold (e.g. by fixing Jira), then OER should become obsoleted.
Is the (hopefully temporary) OER value worth it in the meantime?

**Comment by dagguh** [comment]
> Would it be possible to add the action label to the hover over dialog-thing (a techincal term ;) ) ?

The dot represents a part of `HardwareTestResult`. That means multiple repeats with possibly different worst actions. We could list all (usually 2, sometimes 4) of them. Updating `HardwareTestResult` itself should be trivial. Piping the extra info into the chart is more challenging, because I didn't find a straightforward way to label `chart.js` datasets. It looks like it only contains of a list of X/Y points. Perhaps we could use the [tick callback](https://www.chartjs.org/docs/latest/axes/labelling.html#creating-custom-tick-formats) if we maintain a labelling map.

**Review by atl-mk — APPROVED** [review]

**Review by kema2000 — APPROVED** [review]

**Comment by dagguh** [comment]
I was concerned about the 5% threshold, because it *might* look less marketable.
On a second thought, it's *more* marketable in other ways.
JSW 7.2.0 shows how adding more nodes helps diffuse the errors - DC value.
JSW 7.13.0 shows how much better it is 7.2.0 - upgrade value.

**Comment by dagguh** [comment]
WRT "which action type was the worst?"
We didn't know that either with the overall error rate. But the question is now provoked by the metric. So I'll add this info to the exploration table.

**Comment by dagguh** [comment]
Here's [a build](https://server-gdn-bamboo.internal.atlassian.com/browse/QUICK-JHWR44-1), which should include the most broken action in the exploration table.

**Comment by dagguh** [comment]
I was concerned about breaking symmetry between Apdex and error rate.
Is it ok to leave Apdex action-agnostic and error rate action-specific?
Yes, it is.
Because they have very different consequences for the user.

If Apdex is poor, the workaround is to wait and be frustrated.
If error rate is poor, the workaround is to retry and be blocked in the meantime. You cannot proceed with your work. And if ER is 99%, then even the retries won't help!

That's why we *should* be more strict about errors than about Apdex.

**Comment by dagguh** [comment]
The build has finished. We can see that the biggest offender is consistently "View Backlog".
<details>
<summary>extract from <a href="https://quicksilver-jhwr-cache-ireland.s3-eu-west-1.amazonaws.com/QUICK-240-redo-jsw-7.13.0/xl-jsw-7.13.0/exploration-table.csv?versionId=cqstDcLwUf_MykH4qqTITSooybnOspEA">exploration table</a></summary>
<p>

jira | jira nodes | database | max action error [%] | most errored action
-- | -- | -- | -- | --
c5.2xlarge | 1 | m4.4xlarge | 100.0 | Project Summary
c5.2xlarge | 2 | m4.4xlarge | 100.0 | View Board
c5.2xlarge | 3 | m4.4xlarge | 100.0 | Create Issue
c5.2xlarge | 4 | m4.4xlarge | 100.0 | Browse Boards
c5.2xlarge | 5 | m4.4xlarge | - | -
c5.2xlarge | 6 | m4.4xlarge | - | -
c5.2xlarge | 7 | m4.4xlarge | - | -
c5.2xlarge | 8 | m4.4xlarge | - | -
c5.2xlarge | 9 | m4.4xlarge | - | -
c5.2xlarge | 10 | m4.4xlarge | - | -
c5.2xlarge | 11 | m4.4xlarge | - | -
c5.2xlarge | 12 | m4.4xlarge | - | -
c5.2xlarge | 13 | m4.4xlarge | - | -
c5.2xlarge | 14 | m4.4xlarge | - | -
c5.2xlarge | 15 | m4.4xlarge | - | -
c5.2xlarge | 16 | m4.4xlarge | - | -
c5.4xlarge | 1 | m4.4xlarge | 100.0 | View Board
c5.4xlarge | 2 | m4.4xlarge | 47.94871794871795 | Search with JQL
c5.4xlarge | 3 | m4.4xlarge | 4.21792618629174 | View Backlog
c5.4xlarge | 4 | m4.4xlarge | 5.5762081784386615 | View Backlog
c5.4xlarge | 5 | m4.4xlarge | 4.1450777202072535 | View Backlog
c5.4xlarge | 6 | m4.4xlarge | 4.399524375743163 | View Backlog
c5.4xlarge | 7 | m4.4xlarge | 3.7513397642015005 | View Backlog
c5.4xlarge | 8 | m4.4xlarge | 4.037610619469026 | View Backlog
c5.4xlarge | 9 | m4.4xlarge | - | -
c5.4xlarge | 10 | m4.4xlarge | - | -
c5.4xlarge | 11 | m4.4xlarge | - | -
c5.4xlarge | 12 | m4.4xlarge | - | -
c5.4xlarge | 13 | m4.4xlarge | - | -
c5.4xlarge | 14 | m4.4xlarge | - | -
c5.4xlarge | 15 | m4.4xlarge | - | -
c5.4xlarge | 16 | m4.4xlarge | - | -
c4.8xlarge | 1 | m4.4xlarge | 100.0 | Create Issue
c4.8xlarge | 2 | m4.4xlarge | 80.0 | View Backlog
c4.8xlarge | 3 | m4.4xlarge | 14.723203769140165 | Search with JQL
c4.8xlarge | 4 | m4.4xlarge | 4.592901878914405 | View Backlog
c4.8xlarge | 5 | m4.4xlarge | 5.924339757316202 | View Backlog
c4.8xlarge | 6 | m4.4xlarge | 6.1547836684948205 | View Backlog
c4.8xlarge | 7 | m4.4xlarge | 5.382293762575452 | View Backlog
c4.8xlarge | 8 | m4.4xlarge | 5.1406401551891365 | View Backlog
c4.8xlarge | 9 | m4.4xlarge | 4.681933842239186 | View Backlog
c4.8xlarge | 10 | m4.4xlarge | 6.6492829204693615 | View Backlog
c4.8xlarge | 11 | m4.4xlarge | - | -
c4.8xlarge | 12 | m4.4xlarge | - | -
c4.8xlarge | 13 | m4.4xlarge | - | -
c4.8xlarge | 14 | m4.4xlarge | - | -
c4.8xlarge | 15 | m4.4xlarge | - | -
c4.8xlarge | 16 | m4.4xlarge | - | -
c5.9xlarge | 1 | m4.4xlarge | 100.0 | View Backlog
c5.9xlarge | 2 | m4.4xlarge | 24.734231041814315 | Search with JQL
c5.9xlarge | 3 | m4.4xlarge | 5.859375 | View Backlog
c5.9xlarge | 4 | m4.4xlarge | 4.722366372599896 | View Backlog
c5.9xlarge | 5 | m4.4xlarge | 3.595890410958904 | View Backlog
c5.9xlarge | 6 | m4.4xlarge | 3.5413153456998314 | View Backlog
c5.9xlarge | 7 | m4.4xlarge | 3.2329988851727984 | View Backlog
c5.9xlarge | 8 | m4.4xlarge | - | -
c5.9xlarge | 9 | m4.4xlarge | - | -
c5.9xlarge | 10 | m4.4xlarge | - | -
c5.9xlarge | 11 | m4.4xlarge | - | -
c5.9xlarge | 12 | m4.4xlarge | - | -
c5.9xlarge | 13 | m4.4xlarge | - | -
c5.9xlarge | 14 | m4.4xlarge | - | -
c5.9xlarge | 15 | m4.4xlarge | - | -
c5.9xlarge | 16 | m4.4xlarge | - | -
c5.18xlarge | 1 | m4.4xlarge | 100.0 | View Backlog
c5.18xlarge | 2 | m4.4xlarge | 29.282700421940927 | Search with JQL
c5.18xlarge | 3 | m4.4xlarge | 5.057251908396947 | View Backlog
c5.18xlarge | 4 | m4.4xlarge | 5.145278450363196 | View Backlog
c5.18xlarge | 5 | m4.4xlarge | 4.7841913676547065 | View Backlog
c5.18xlarge | 6 | m4.4xlarge | 4.1281912004345465 | View Backlog
c5.18xlarge | 7 | m4.4xlarge | 3.9344262295081966 | View Backlog
c5.18xlarge | 8 | m4.4xlarge | 3.2054951345163136 | View Backlog
c5.18xlarge | 9 | m4.4xlarge | 4.073199527744983 | View Backlog
c5.18xlarge | 10 | m4.4xlarge | - | -
c5.18xlarge | 11 | m4.4xlarge | - | -
c5.18xlarge | 12 | m4.4xlarge | - | -
c5.18xlarge | 13 | m4.4xlarge | - | -
c5.18xlarge | 14 | m4.4xlarge | - | -
c5.18xlarge | 15 | m4.4xlarge | - | -
c5.18xlarge | 16 | m4.4xlarge | - | -
</p>
</details>

We can raise the problem to the JSW team, hopefully get a public ticket and link it in the docs.
WDYT @atlassian/quicksilver?

**Comment by dagguh** [comment]
In a discussion with @asanzrivas, we tend to agree that showing the overall picture is worth it, so in short:
> Is the (hopefully temporary) OER value worth it in the meantime?

Yes.

Now we have a choice between overall error rate (OER) and average action error rate (AAER).
OER has heterogenous actions - ie. View Issue is much more important than Browse Projects.
AAER has homogenous actions - ie. all actions are equally important.
Traditionally we used OER to reflect the frustration frequency of users.

**Comment by dagguh** [comment]
I pushed in both MAER and OER at the same time. I also pushed the MAER threshold.
Here's a [build](https://server-gdn-bamboo.internal.atlassian.com/browse/QUICK-JHWR45-1).

**Pull request merged**